### PR TITLE
extract optimizer loop

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7938

Extract an OptimizerLoop from the existing TrainingBatchLoop. This cleanly separates the code path for automatic and manual optimization. Automatic optimization now completely lives inside the OptimizerLoop.

### Previously:
(details left out)

```python
if lightning_module.automatic_optimization:
    for opt_idx, opt in enumerate(trainer.optimizers):
        self._run_optimization(batch, opt_idx, opt)
else:
     self._run_optimization(batch, 0, None)

```
The `run_optimization` had to handle both code paths for automatic and manual optimization, making it hard to follow, debug, and extend.

### Now:
(details left out)

```python
if lightning_module.automatic_optimization:
    self.optimizer.run(batch, optimizers)
else:
     self._run_manual_optimization(batch)
```
The code path branches here cleanly. The new OptimizerLoop handles automatic optimization completely and the manual optimization path is kept isolated. This greatly improves readability, debugging, and extensibility for loop customization.

A follow-up PR will remove some remaining traces of mixed interfaces, we are holding it out from this PR to make it more manageable to review.

Blocked by:
- #9188
- #9192
- #9198

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 1]
```diff
@@ -240,51 +204,6 @@ def _make_step_fn(self, split_batch: Any, batch_idx: int, opt_idx: int, hiddens:
         """Build the step function that runs the `training_step` and processes its output."""
         return partial(self._training_step, split_batch, batch_idx, opt_idx, hiddens)
 
-    def _make_zero_grad_fn(self, batch_idx: int, opt_idx: int, optimizer: Optimizer) -> Optional[Callable[[], None]]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:243]: reviewer: this removed block is all specific to automatic optimization and gets moved to the OptimizerLoop

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 2]
```diff
@@ -334,97 +256,7 @@ def _training_step(
...
-        closure_loss = None
-        loss = None
-        if self.trainer.lightning_module.automatic_optimization:
-            # accumulate loss. if accumulate_grad_batches==1, no effect
-            closure_loss = result_collection.minimize / self.trainer.accumulate_grad_batches
-            # the loss will get scaled for amp. avoid any modifications to it
-            loss = closure_loss.detach().clone()
-        return AttributeDict(closure_loss=closure_loss, loss=loss, result_collection=result_collection)
-
-    def _optimizer_step(
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:346]: reviewer: these removed methods below are all specific to automatic optimization and get moved to the OptimizerLoop

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 3]
```diff
@@ -334,97 +256,7 @@ def _training_step(
             if result_collection is None:
                 return
 
-        closure_loss = None
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:337]: reviewer: this removed block is all specific to automatic optimization and gets moved to the OptimizerLoop

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 4]
```diff
@@ -441,47 +273,6 @@ def _tbptt_split_batch(self, batch: Any) -> List[Any]:
...
-            model = self.trainer.lightning_module
-            model.toggle_optimizer(optimizer, opt_idx)
-
-    def _run_optimization_end(self, opt_idx: int) -> None:
-        if self.trainer.lightning_module.automatic_optimization and len(self.trainer.optimizers) > 1:
-            model = self.trainer.lightning_module
-            model.untoggle_optimizer(opt_idx)
-
-    @contextmanager
-    def block_ddp_sync_behaviour(self, should_block_sync: bool = False) -> Generator[None, None, None]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:464]: reviewer: this got split out into a separate PR, see #9192

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 5]
```diff
@@ -498,13 +289,6 @@ def backward(
             opt_idx: Index of the current optimizer being used. ``None`` if using manual optimization.
         """
         self.trainer.accelerator.backward(loss, optimizer, opt_idx, *args, **kwargs)
-
-        if not self.trainer.fit_loop.should_accumulate():
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:502]: reviewer: this removed block is all specific to automatic optimization and gets moved to the OptimizerLoop

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 6]
```diff
@@ -540,46 +324,3 @@ def get_active_optimizers(self, batch_idx: Optional[int] = None) -> List[Tuple[i
         # find optimzier index by looking for the first {item > current_place} in the cumsum list
         opt_idx = int(np.argmax(self.optimizer_freq_cumsum > current_place_in_loop))
         return [(opt_idx, self.trainer.optimizers[opt_idx])]
-
-    def _build_kwargs(self, batch: Any, batch_idx: int, opt_idx: int, hiddens: Optional[Tensor]) -> Dict[str, Any]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:544]: reviewer: moved to utilities to share code with OptimizerLoop, split out in #9198

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+                return
+
+        # accumulate loss. if accumulate_grad_batches==1, no effect
+        closure_loss = result_collection.minimize / self.trainer.accumulate_grad_batches
+        # the loss will get scaled for amp. avoid any modifications to it
+        loss = closure_loss.detach().clone()
+        return AttributeDict(closure_loss=closure_loss, loss=loss, result_collection=result_collection)
+
+    # TODO: this is duplicated
+    def _process_closure_result(self, opt_closure_result: Optional[ClosureResult]) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:365]: reviewer: this will be removed entirely in #9189

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 8]
```diff
@@ -161,68 +150,37 @@ def _run_optimization(
...
-        # ------------------------------
-        # gradient update with accumulated gradients
-        else:
-            if self.trainer.lightning_module.automatic_optimization:
-                self._optimizer_step(optimizer, opt_idx, batch_idx, closure)
-            else:
-                closure()
-
+        closure = self._make_closure(split_batch, batch_idx, self._hiddens)
+        closure()
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:161]: eliminating the closure here for manual optimization will be a direct follow up to this PR

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 9]
```diff
@@ -53,13 +49,12 @@ def __init__(self) -> None:
         self.running_loss: TensorRunningAccum = TensorRunningAccum(window_length=20)
         # the current split index when the batch gets split into chunks in truncated backprop through time
         self.split_idx: Optional[int] = None
-        self.optim_progress = OptimizationProgress()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:53]: Would this work with manual optimization tracking ?

File: pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 10]
```diff
@@ -216,13 +216,13 @@ def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningMo
             # performing only one pass over the train data-loader to compute activation statistics
             # Therefore, we will virtually increase `num_training_batches` by 1 and skip backward.
             trainer.num_training_batches += 1
-            trainer.fit_loop._skip_backward = True
+            trainer.fit_loop.epoch_loop.batch_loop.optimizer_loop._skip_backward = True
```
Comment (carmocca) [inline pytorch_lightning/callbacks/stochastic_weight_avg.py:219]: Why not use the property in `fit_loop` for it?

```suggestion
            trainer.fit_loop._skip_backward = True
```

Otherwise remove the property itself

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 11]
```diff
@@ -132,17 +126,12 @@ def advance(self, batch, batch_idx):
...
-                    if opt_idx < self.optim_progress.optimizer_idx:
-                        continue
-
-                self.optim_progress.optimizer_idx = opt_idx
-
-                result = self._run_optimization(batch_idx, split_batch, opt_idx, optimizer)
-                if result:
-                    self.batch_outputs[opt_idx].append(copy(result.result_collection))
+            # in automatic optimization, hand over execution to the OptimizerLoop
+            optimizers = [optimizer for _, optimizer in self.get_active_optimizers(batch_idx)]
```
Comment (tchaton) [inline pytorch_lightning/loops/batch/training_batch_loop.py:130]: Should the `OptimizerLoop` takes care of active `optimizers` creation and this would be moved to `on_run_start` and used for self.done ?

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+        raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
+
+    def reset(self) -> None:
+        if not self.restarting:
+            self.optim_progress.optimizer_idx = 0
+        self.outputs = [[] for _ in range(len(self.trainer.optimizers))]
+
+    def on_run_start(self, batch: Any, hiddens: Any, optimizers: List[Optimizer], batch_idx: int) -> None:
+        self._batch_idx = batch_idx
+        self._optimizers = optimizers.copy()
```
Comment (tchaton) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:71]: Why are you copying the optimizers ?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 13]
```diff
@@ -132,17 +126,12 @@ def advance(self, batch, batch_idx):
...
+            batch_outputs, self._hiddens = self.optimizer_loop.run(split_batch, self._hiddens, optimizers, batch_idx)
+            # combine outputs from each optimizer
+            for k in range(len(batch_outputs)):
+                self.batch_outputs[k].extend(batch_outputs[k])
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:134]: Why do we extend now instead of append as in the previous version?

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py:161]: Add a TODO then :)

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+
+        if not self.trainer.fit_loop.should_accumulate():
+            # track gradients
+            grad_norm_dict = self._track_and_norm_grad(optimizer=optimizer)
+            if grad_norm_dict:
+                self.trainer.lightning_module._current_fx_name = "on_after_backward"
+                self.trainer.lightning_module.log_grad_norm(grad_norm_dict)
+        return loss
+
+    def _run_optimization(
```
Comment (tchaton) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:119]: Do you think this could also become a utility ?

```py
def run_optimization(
    closure: LightningClosure,
    should_accumulate,
    optimizer,
    opt_idx,
    trainer,
    optimizer_step,
) -> Optional[ClosureResult]:
    """Runs closure (train step + backward) together with optimization if necessary.
    Args:
        batch_idx: the index of the current batch
        split_batch: the current tbptt split of the whole batch
        opt_idx: the index of the current optimizer or `None` in case of manual optimization
        optimizer: the current optimizer or `None` in case of manual optimization
    """
    run_optimization_start(opt_idx, optimizer)

    if should_accumulate:
        # For gradient accumulation

        # -------------------
        # calculate loss (train step + train step end)
        # -------------------
        # automatic_optimization=True: perform ddp sync only when performing optimizer_step
        with _block_parallel_sync_behavior(trainer, block=True):
            closure()

    # ------------------------------
    # BACKWARD PASS
    # ------------------------------
    # gradient update with accumulated gradients
    else:
        optimizer_step(optimizer, opt_idx, batch_idx, closure)

    result = closure.get_result()

    # untoggle model params
    run_optimization_end(opt_idx)
    return result
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+    def __init__(self):
+        super().__init__()
+        # TODO: use default dict here to simplify logic in loop
+        self.outputs: _OUTPUTS_TYPE = []
+        self.optim_progress: OptimizationProgress = OptimizationProgress()
+
+        self._skip_backward: bool = False
+        self._batch_idx: Optional[int] = None
+        self._optimizers: Optional[List[Optimizer]] = None
+        self._hiddens: Optional[Any] = None
```
Comment (carmocca) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:54]: Does this need to keep an attribute for `_hiddens` in this loop? We already have one in the batch loop. And that one is the one dealing with split indices etc

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    def done(self) -> bool:
+        """Returns ``True`` when the last optimizer in the sequence has run."""
+        return self.optim_progress.optimizer_idx >= len(self._optimizers)
+
+    def connect(self, **kwargs: "Loop") -> None:
+        raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
+
+    def reset(self) -> None:
+        if not self.restarting:
+            self.optim_progress.optimizer_idx = 0
```
Comment (carmocca) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:66]: Is this necessary?

The `reset_on_epoch` call in the parent loop will set this already

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:71]: ah thanks, not needed! I was popping from the list originally but didn't work out so nicely. will remove!

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:130]: for the other loops, we are also creating the iterable outside and pass it to run()
e.g. in fit loop where we create the dataloader and then pass it to the epoch loop. 

I was following the same pattern here.

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 12 above)
Comment (carmocca) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:71]: Why do we need to `copy`?

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 14 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:119]: you are right, I think that is useful!

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:134]: there is an append inside the optimizer loop to it's own outputs list. 
here we need to extend to make it equivalent to the previous case where we had a single outputs list for the combined loop.

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:71]: we don't anymore :) removed.

the initial approach I had was to pop from the list every advance.

Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py]: removed!

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 15 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:54]: yes, I think it can't be avoided because hiddens needs to pass through the training step and then be returned back again to the parent loop. I need to somehow return it in the `on_run_end()` method.

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 16 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:66]: yes this is needed. 
with every new batch we want to rerun over the optimizers from the start. And in case we are restarting, we want to resume from the last optimizer we failed at.

I think the reset method you mentioned is for resetting the counters of steps for the epoch. Here we are concerned about the current optimizer idx

File: pytorch_lightning/callbacks/stochastic_weight_avg.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/stochastic_weight_avg.py:219]: i was anticipating this comment.
on the one hand skip_backward has nothing to do with the fit loop
on the other hand we don't want to reach so deep into the loop stack, it  becomes fragile when opening up the loops for customization.

related: #9220

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:53]: no, we are currently not updating optim_progress for manual optimization

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 14 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:119]: added this as a direct follow up #7938 as it is easier to do once this PR is merged

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:134]: happy to hear suggestions for how to simplify outputs handling across loops, but definitely out of scope for this PR.

Comment (SeanNaren) [reply pytorch_lightning/loops/batch/training_batch_loop.py:134]: This is fine with me!

**Review by SeanNaren — APPROVED** [review]
LGTM, great work!

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 17]
```diff
@@ -416,25 +246,7 @@ def _tbptt_split_batch(self, batch: Any) -> List[Any]:
...
-        if self.trainer.lightning_module.automatic_optimization and len(self.trainer.optimizers) > 1:
-            model = self.trainer.lightning_module
-            model.toggle_optimizer(optimizer, opt_idx)
-
-    def _run_optimization_end(self, opt_idx: int) -> None:
-        if self.trainer.lightning_module.automatic_optimization and len(self.trainer.optimizers) > 1:
-            model = self.trainer.lightning_module
-            model.untoggle_optimizer(opt_idx)
-
+    # TODO: remove this method and update tests
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:249]: Is there a reason why you are not doing this in this PR?

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 16 above)
Comment (carmocca) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:66]: But perhaps we should remove this line as it's done here

https://github.com/PyTorchLightning/pytorch-lightning/blob/f9994d456cb264f8d66002eae6d7d51bd1ecc94d/pytorch_lightning/trainer/progress.py#L196

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 17 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:249]: to keep the PR small and reviewer happy :)

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 16 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:66]: yes true, resetting optimizer_idx = 0 does not really relate to the epoch end

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 17 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:249]: #9265

**Review by tchaton — APPROVED** [review]
Great work !

File: pytorch_lightning/loops/optimizer/optimizer_loop.py (hunk 15 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py:54]: Do you agree? Or how would we avoid it?

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:130]: I would prefer to have the splitting of the batch inside the loop that corresponds to the TBPTT, which is TrainingBatchLoop. This way the loop can be customized and the parent loop does not need knowledge about whether the child loop is requiring a split batch or not.
